crates/walrus-service/src/node/node_recovery.rs

```diff
@@ -217,22 +230,26 @@ impl NodeRecoveryHandler {
     /// Restarts any in progress recovery.
     pub async fn restart_recovery(&self) -> Result<(), TypedStoreError> {
         if let NodeStatus::RecoveryInProgress(recovering_epoch) = self.node.storage.node_status()? {
-            if recovering_epoch == self.node.current_epoch() {
-                // The `latest_event_epoch` is still set to `0` at this point.
-                self.node
-                    .latest_event_epoch
-                    .store(recovering_epoch, Ordering::SeqCst);
-                return self.start_node_recovery(self.node.current_epoch()).await;
-            } else {
-                assert!(recovering_epoch < self.node.current_epoch());
-                tracing::warn!(
-                    recovering_epoch,
-                    current_epoch = self.node.current_epoch(),
-                    "recovery epoch mismatch; skip recovery restart; next epoch change start event \
-                    will bring node to the latest state"
-                );
-                self.node.set_node_status(NodeStatus::RecoveryCatchUp)?;
+            // Wait until the latest_event_epoch watcher is set to Some(epoch)
+            let mut watcher = self.node.latest_event_epoch_watcher();
+
+            tracing::info!("waiting for latest event epoch to be set to restart node recovery");
+
+            // Wait for the node.latest_event_epoch is initialized, so that the node recovery
+            // task can start.
+            while watcher.borrow().is_none() {
+                watcher
+                    .changed()
+                    .await
+                    .expect("watcher should not be closed");
             }
```

Comment:
Suggestion: Make the `current_event_epoch` function `async` and put this into it's body.